### PR TITLE
Kamu 129 implement event details support for set polling source event

### PR DESCRIPTION
**Pull request by dmitriy-borzenko**

### CODE REVIEW

File: src/app/api/gql/fragments/events/fragment-set-polling-source.graphql [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+      args
+      env {
+        name
+        value
+      }
+    }
+  }
+  read {
+    ... on ReadStepCsv {
+      schema
```
Comment (zaychenko-sergei) [inline src/app/api/gql/fragments/events/fragment-set-polling-source.graphql:43]: It looks `schema` is in every branch, can we move it 1 level up?

File: src/app/components/toggle/toggle.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { ComponentFixture, TestBed } from "@angular/core/testing";
+
+import { ToggleComponent } from "./toggle.component";
+
+describe("ToggleComponent", () => {
```
Comment (zaychenko-sergei) [inline src/app/components/toggle/toggle.component.spec.ts:5]: Need more tests - the clicks are not covered

File: src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html [hunk 3]
```diff
@@ -0,0 +1,202 @@
+<div class="d-flex justify-content-end">
+    <span class="me-2">Show as yaml</span>
+    <app-toggle
+        [isToggled]="isYamlView"
+        (toggled)="onToggleView($event)"
+    ></app-toggle>
+</div>
+
+<div *ngIf="isYamlView" class="layout lh-1 font-monospace mt-4">
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html:9]: We need to think of something more universal. There would be many kinds of events, and implementing the YAML view directly each time would be annoying. Instead, what I'd like to see is a universal component that can show some YAML structures based on the model.

So:
 - some object model describing YAML sections (something simple, tree-like structure)
 - some universal visual component (or components) that can render this model
 - a builder of the YAML model, that would produce it based on the input event structures, and depending on their types/data

File: src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    </ng-container>
+    <p class="color-key">read:</p>
+    <p>
+        &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;kind:
+        <span class="value-color">{{ event.read.__typename }}</span>
+    </p>
+    <ng-container *ngIf="isReadStepCsv">
+        <p>
+            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;separator:
+            <span class="value-color"> {{ readStepCsv.separator }}</span>
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html:44]: The separator symbol is almost invislble on the form. Maybe it should be in quotes, or with an extra explanation text, i.e.:

  "," (comma)

File: src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    <ng-container *ngIf="event.preprocess">
+        <p class="color-key">preprocess:</p>
+        <p>
+            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;kind:
+            <span class="value-color"> {{ event.preprocess.__typename }}</span>
+        </p>
+
+        <p>
+            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;engine:
+            <span class="value-color"> {{ event.preprocess.engine }}</span>
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html:74]: Would be good to display engines by full names:
 - Apache Spark
 - Apacke Flink
 
 or even better, with logos:
 
image

image

File: src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+                        class="value-color"
+                    >
+                        {{ query.alias ? query.alias : "none" }}</span
+                    >
+                </p>
+                <p>
+                    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; query:<span
+                        class="value-color lh-sm"
+                    >
+                        {{ query.query }}</span
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html:96]: Query should be better formatted with SQL syntax highlighting.
Maybe we need a read-only Monaco component, or something simpler, as we don't need to edit.
This probably is out of scope for this ticket, still would be good to think how we are going to implement this.

File: src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        <p>
+            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;header:
+            <span class="value-color"> {{ readStepCsv.header }}</span>
+        </p>
+    </ng-container>
+    <p class="color-key">merge:</p>
+    <ng-container *ngIf="isMergeStrategyLedger">
+        <p>
+            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;kind:
+            <span class="value-color"> {{ event.merge.__typename }}</span>
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html:55]: Can we encapsulate all these `__typename` low level stuff in the component code behind properties?

**Review by zaychenko-sergei — CHANGES_REQUESTED** [review]

File: src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html (hunk 4 above)
Comment (zaychenko-sergei) [reply src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html:44]: Or maybe icon, or maybe larger font.
I don't restrict the mechanism how it can be improved, just make it clearly visible.

File: src/app/api/gql/fragments/events/fragment-set-polling-source.graphql (hunk 1 above)
Comment (dmitriy-borzenko) [reply src/app/api/gql/fragments/events/fragment-set-polling-source.graphql:43]: No, we can't move this field, because it's a union.

File: src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html (hunk 4 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html:44]: Fixed. Added larger font and inline quotation.

File: src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html (hunk 6 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html:96]: Added a read-only Monaco component

**Review by zaychenko-sergei — CHANGES_REQUESTED** [review]
Is Flink icon transparent?

File: src/app/common/data.helpers.ts [hunk 8]
```diff
@@ -27,6 +27,24 @@ export class DataHelpers {
...
+        name: string;
+        url_logo: string;
+    } {
+        switch (name) {
+            case "flink":
+                return {
+                    name: "Apache Flink",
+                    url_logo: "assets/images/apache-flink.png",
+                };
+            default:
```
[the remark this answers is not in the document]
Comment (dmitriy-borzenko) [reply src/app/common/data.helpers.ts:40]: Fixed

File: src/app/common/data.helpers.ts [hunk 9]
```diff
@@ -27,6 +27,24 @@ export class DataHelpers {
     public static readonly BLOCK_DESCRIBE_SET_ATTACHMENTS =
         "Attachments updated";
 
+    public static descriptionForEngine(name: string): {
+        name: string;
+        url_logo: string;
```
[the remark this answers is not in the document]
Comment (dmitriy-borzenko) [reply src/app/common/data.helpers.ts:32]: Added interface

File: src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        language: "sql",
+        contextmenu: false,
+        wordWrap: "on",
+        readOnly: true,
+        renderLineHighlight: "none",
+        lineNumbers: "off",
+        minimap: {
+            enabled: false,
+        },
+    };
```
[the remark this answers is not in the document]
Comment (dmitriy-borzenko) [reply src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.ts:33]: Moved this config into separate file.

File: src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+    public get mergeStrategyLedger(): MergeStrategyLedger {
+        return this.event.merge as MergeStrategyLedger;
+    }
+
+    public get isMergeStrategyLedger(): boolean {
+        return this.event.merge.__typename === "MergeStrategyLedger";
+    }
+
+    public descriptionMergeStrategy(type: string | undefined): string {
```
[the remark this answers is not in the document]
Comment (dmitriy-borzenko) [reply src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.ts:67]: Yes, added icons for MergeStrategy types

File: src/app/dataset-block/metadata-block/components/event-details/components/common/base-property/base-property.component.ts [hunk 12]
```diff
@@ -0,0 +1,9 @@
+import { Component, Input } from "@angular/core";
+
+@Component({
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/common/base-property/base-property.component.ts:3]: @Component annotation is normally not required for abstract classes. It's enough that the concrete subclasses define it

File: src/app/dataset-block/metadata-block/components/event-details/components/common/separator-property/separator-property.component.html [hunk 13]
```diff
@@ -0,0 +1,3 @@
+<q class="fs-5">
+    <span class="fs-4">{{ data ? data : defaultSeparator }}</span>
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/common/separator-property/separator-property.component.html:2]: I expected you'll use the `class` input here

File: src/app/dataset-block/metadata-block/components/event-details/supported.events.ts [hunk 14]
```diff
@@ -10,3 +10,8 @@ export enum SupportedEvents {
     SetVocab = "SetVocab",
     SetWatermark = "SetWatermark",
 }
+
+export interface LogoInfo {
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/supported.events.ts:14]: It's either a wrong file to define `LogoInfo`, or the name of the file is wrong. Currently it's suggesting we focus on supported events. Logo is not about events.

File: src/app/dataset-block/metadata-block/components/event-details/components/common/base-property/base-property.component.ts [hunk 15]
```diff
@@ -0,0 +1,9 @@
...
+    template: "",
+})
+export abstract class BasePropertyComponent {
+    @Input() public data: unknown;
+    @Input() public class?: string;
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/common/base-property/base-property.component.ts:8]: Logically, `class` property rather belongs to `SimplePropertyComponent` than here.

File: src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.ts [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+})
+export class SetPollingSourceEventComponent implements OnInit, AfterViewInit {
+    @Input() public event: SetPollingSource;
+    @Input() public datasetInfo: DatasetInfo;
+    @ViewChildren("readStepContainer", { read: ViewContainerRef })
+    readStepContainer: QueryList<ViewContainerRef>;
+    public isYamlView = false;
+    public sqlEditorOptions = sqlEditorOptionsForEvents;
+    public readStepCsvData: EventRow[];
+    private componentRef: ComponentRef<BasePropertyComponent>;
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.ts:42]: Logicaly, it's rather a local temporary variable within `ngAfterViewInit`, than a long-lived property

**Review by zaychenko-sergei — CHANGES_REQUESTED** [review]
When clicking on "show as yaml", and back, the data disappears

File: src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (zaychenko-sergei) [reply src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.ts]: Does this record have any type in Monaco world ?

File: src/app/dataset-block/metadata-block/components/event-details/components/common/editor-property/editor-property.component.ts [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+    selector: "app-editor-property",
+    templateUrl: "./editor-property.component.html",
+    styleUrls: ["./editor-property.component.sass"],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class EditorPropertyComponent extends BasePropertyComponent {
+    public sqlEditorOptions = sqlEditorOptionsForEvents;
+    public test =
+        'SELECT CAST(UNIX_TIMESTAMP(Reported_Date, "yyyy-MM-dd") as TIMESTAMP) as reported_date, Classification_Reported as classification, ROW_NUMBER() OVER (ORDER BY (Reported_Date, HA)) as id, ha, sex, age_group FROM input';
+
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/common/editor-property/editor-property.component.ts:16]: Remove debug code

File: src/app/dataset-block/metadata-block/components/event-details/components/common/simple-property/simple-property.component.html [hunk 18]
```diff
@@ -0,0 +1 @@
+<span [class]="class">{{ data }}</span>
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/common/simple-property/simple-property.component.html:1]: `class` is optional in the component. Here it is used without checks. If we always have a class, change the the property or have a default value maybe.

File: src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+            <div class="mx-4">
+                <ng-container *ngFor="let row of section.rows">
+                    <app-block-row-data
+                        *ngIf="row.label !== queriesLabel"
+                        [tooltip]="row.tooltip"
+                        [label]="row.label"
+                    >
+                        <ng-template #container></ng-template>
+                    </app-block-row-data>
+                    <ng-container *ngIf="row.label === queriesLabel">
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/set-polling-source-event/set-polling-source-event.component.html:123]: Instead of detecting the case by text of the label, introduce an explicit boolean field in the row interface - something like `separate_row_for_value: bool`

File: src/app/dataset-block/metadata-block/components/event-details/factory.events.ts [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+
+enum SetPollingSourceProperty {
+    READ = "read",
+    FETCH = "fetch",
+    MERGE = "merge",
+    PREPROCESS = "preprocess",
+    PREPARE = "prepare",
+}
+
+abstract class EventRowFactory<TEvent> {
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/factory.events.ts:32]: Let's have good naming here:
 - this is becoming rather a builder than a factory
 - we are building sections with rows, not just rows
So, perhaps, `EventSectionBuilder`.

File: src/app/dataset-block/metadata-block/components/event-details/factory.events.ts [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+                    Object.entries(event[property]).forEach(([key, value]) => {
+                        if (value && key !== "__typename") {
+                            rows.push({
+                                ...SET_POLLING_SOURCE_DESCRIPTORS[
+                                    `SetPollingSource.${event[property].__typename}.${key}`
+                                ],
+                                value: value,
+                            });
+                        }
+                    });
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/factory.events.ts:52]: This block looks identical in all 5 event steps. Make it a subprogram.

File: src/app/dataset-block/metadata-block/components/event-details/factory.events.ts [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+    public abstract buildEventRows(event: TEvent): EventSection[];
+}
+
+class SetPollingSourceEventRowFactory extends EventRowFactory<SetPollingSource> {
+    public buildEventRows(event: SetPollingSource): EventSection[] {
+        const result: EventSection[] = [];
+        let rows: EventRow[] = [];
+        Object.entries(event).forEach(([property, data]) => {
+            if (data && property !== "__typename") {
+                if (property === SetPollingSourceProperty.READ) {
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/factory.events.ts:42]: I think that `switch..case` statement exists in TypeScript too :-)

File: src/app/dataset-block/metadata-block/components/event-details/factory.events.ts [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+abstract class EventRowFactory<TEvent> {
+    public abstract buildEventRows(event: TEvent): EventSection[];
+}
+
+class SetPollingSourceEventRowFactory extends EventRowFactory<SetPollingSource> {
+    public buildEventRows(event: SetPollingSource): EventSection[] {
+        const result: EventSection[] = [];
+        let rows: EventRow[] = [];
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/factory.events.ts:39]: `rows` is a local container of the section, it should not be defined at this level. It should be re-created in each section

File: src/app/dataset-block/metadata-block/components/event-details/mock.events.ts [hunk 24]
```diff
@@ -25,6 +26,50 @@ export const mockAddData: AddDataEventFragment = {
     },
 };
 
+export const mockSetPollinfSourceEvent: SetPollingSource = {
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/mock.events.ts:29]: pollin**g**

File: src/app/dataset-block/metadata-block/components/event-details/components/common/editor-property/editor-property.component.ts [hunk 25]
```diff
@@ -0,0 +1,17 @@
...
+import { sqlEditorOptionsForEvents } from "../../../config-editor.events";
+import { BasePropertyComponent } from "../base-property/base-property.component";
+
+@Component({
+    selector: "app-editor-property",
+    templateUrl: "./editor-property.component.html",
+    styleUrls: ["./editor-property.component.sass"],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class EditorPropertyComponent extends BasePropertyComponent {
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/common/editor-property/editor-property.component.ts:12]: The name is no longer adequate. We don't use an editor anymore, even a read-only one. Besides, it's specificly showing SQL. So it is an `SqlQueryViewerComponent`

File: src/app/common/table.helper.ts [hunk 26]
```diff
@@ -0,0 +1,9 @@
+import { DataRow, DataSchemaField } from "../interface/dataset.interface";
+
+export const createSchemaFields = (data: DataRow): DataSchemaField[] => {
```
Comment (zaychenko-sergei) [inline src/app/common/table.helper.ts:3]: extractSchemaFieldsFromData

File: src/app/dataset-block/metadata-block/components/event-details/builder.events.ts [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+
+enum SetPollingSourceProperty {
+    READ = "read",
+    FETCH = "fetch",
+    MERGE = "merge",
+    PREPROCESS = "preprocess",
+    PREPARE = "prepare",
+}
+
+const keysForSeparateRow: string[] = ["queries"];
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/builder.events.ts:33]: Make a boolean property in the main descriptors instead, it should be `false` by default.

File: src/app/dataset-block/metadata-block/components/event-details/builder.events.ts [hunk 28]
```diff
@@ -0,0 +1,115 @@
...
+                        const rows: EventRow[] = [];
+                        Object.entries(event[property]).forEach(
+                            ([key, value]) => {
+                                if (value && key !== "__typename") {
+                                    rows.push({
+                                        ...SET_POLLING_SOURCE_DESCRIPTORS[
+                                            `SetPollingSource.${event[property].__typename}.${key}`
+                                        ],
+                                        value: value,
+                                    });
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/builder.events.ts:56]: This internal block that fills the rows should become a subprogram. Reuse it 3 times.

File: src/app/dataset-block/metadata-block/components/event-details/builder.events.ts [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+abstract class EventSectionBuilder<TEvent> {
+    public abstract buildEventRows(event: TEvent): EventSection[];
+}
+
+class SetPollingSourceSectionBuilder extends EventSectionBuilder<SetPollingSource> {
+    public buildEventRows(event: SetPollingSource): EventSection[] {
+        const result: EventSection[] = [];
+        Object.entries(event).forEach(([property, data]) => {
+            if (data && property !== "__typename") {
+                switch (property) {
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/builder.events.ts:44]: A default handler is missing - what if other sections are added

File: src/app/dataset-block/metadata-block/components/event-details/builder.events.ts [hunk 30]
```diff
@@ -0,0 +1,115 @@
...
+                    case SetPollingSourceProperty.READ:
+                    case SetPollingSourceProperty.FETCH: {
+                        const rows: EventRow[] = [];
+                        Object.entries(event[property]).forEach(
+                            ([key, value]) => {
+                                if (value && key !== "__typename") {
+                                    rows.push({
+                                        ...SET_POLLING_SOURCE_DESCRIPTORS[
+                                            `SetPollingSource.${event[property].__typename}.${key}`
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/builder.events.ts:53]: No handling if key is unsupported. Should create a default component with just a string value

File: src/app/dataset-block/metadata-block/components/event-details/components/common/base-property/base-property.component.ts [hunk 31]
```diff
@@ -0,0 +1,7 @@
+import { EnvVar } from "./../../../../../../../api/kamu.graphql.interface";
+import { Injectable, Input } from "@angular/core";
+
+@Injectable()
+export abstract class BasePropertyComponent {
+    @Input() public data: unknown | string[] | EnvVar[];
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/common/base-property/base-property.component.ts:6]: I don't see sense in having this union input. Instead, put a well-typed input in each subclas. No benefits from calling different things with the name `data`.

File: src/app/dataset-block/metadata-block/components/event-details/builder.events.ts [hunk 32]
```diff
@@ -0,0 +1,115 @@
...
+                                        ],
+                                        value: value,
+                                    });
+                                }
+                            },
+                        );
+                        result.push({ title: property, rows });
+                        break;
+                    }
+                    case SetPollingSourceProperty.MERGE: {
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/builder.events.ts:63]: Showing ledger icons broke - please check this code

**Comment by zaychenko-sergei** [comment]
Let's improve test coverage of the builder a bit, and that's enough at this stage!
